### ISSUE TITLE
Great plug-in, however, The ability to directly address the input by name e.g. Select Input "Apple TV", "Disc Player".... would be a much better experience than toggling sequentially through the inputs

### ISSUE THREAD

**Issue by CaptainPic** [EHylands/homebridge-anthemreceiver#6]
**Is your feature request related to a problem? Please describe:**

**Describe the solution you'd like:**

**Describe alternatives you've considered:**

**Additional context:**

**Comment by EHylands** [comment]
Hi @CaptainPic ,
Did you installed the Televison external accessory for your zone?

External television accessories are needed to directly select an input. An extra step is needed to pair the accessory with Homekit.

![image](https://user-images.githubusercontent.com/90003064/202880901-dd070a3d-3f77-4b8a-af7a-913d5eb2171e.png)

![image](https://user-images.githubusercontent.com/90003064/202880902-d477278e-bf91-4fca-bf09-3ec1e2f916ec.jpeg)

![image](https://user-images.githubusercontent.com/90003064/202902911-77fddead-8baa-4c4a-b93a-3de7a8518024.jpeg)

**Comment by EHylands** [comment]
Or would you like the the plugin to create one switch for every input and allow direct input selection ?

**Comment by CaptainPic** [comment]
Yes that's probably the best way  just like a button for each of the inputs

**Comment by EHylands** [comment]
You can install alternate version 0.6.3-beta in Homebridge-UI.
I created a new accessory (Input Selector)

Unfortunatly, there is a bug in iOS 16 and all sub accessories under the main accessory wil have the same name.
If you select each accessory detail view and delete their name, the intended name will be writen as a placeholder. 

Let me know if you have any comments !

**Comment by EHylands** [comment]
![image](https://user-images.githubusercontent.com/90003064/204096005-b0b4941e-e348-468d-9995-25341524b790.png)

**Comment by CaptainPic** [comment]
Hi    Just saw this – let me see if I can get it set up in the next day or two

Thanks


Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by CaptainPic** [comment]
Got it partially working.

Nice the way it auto populated the names I had on my TV and Receiver for inputs, but in the Apple Home App its pretty confusing group of switches and the input names did not carry over there although it looks like you got the labeling in the screen shot you sent me.  It does appear they are mutually exclusive as must be the case since only one input at a time is selectalbe.   See screen shot below.

I haven't had a chance to try to clean up the naming and I don't believe there is any flexibility in Apple home to change the icons.  Would be nice to have buttons and show the volume mute controls in the speakers section.  I expect these are all Apple Home app restrictions and I havent gone too deep into the Homekit add on apps.
'
All that said I'm more interested in using the input selectors via Siri than the Home App UI.  As long as they can be incorporated into an Siri Short cut and automation this might work for me.  Where I'm going is to say Hey Siri turn on Apple tV or Hey Siri turn on the disc player...and have everything turned on with the right input settings.   Volume would be nice but I know there is the complication with the AI's reserving the phrase for volume to control the smart speaker itself.

Give me a couple days and I'll dive into this a bit more and provide more comments.

Thanks

Jack

[cid:81aa1689-15a8-44d0-805d-932a1c0d0434]
________________________________
From: Jack Denenberg ***@***.***>
Sent: Tuesday, November 29, 2022 11:29 AM
To: EHylands/homebridge-anthemreceiver ***@***.***>
Subject: RE: [EHylands/homebridge-anthemreceiver] Great plug-in, however, The ability to directly address the input by name e.g. Select Input "Apple TV", "Disc Player".... would be a much better experience than toggling sequentially through the inputs (Issue #6)


Hi    Just saw this – let me see if I can get it set up in the next day or two



Thanks





Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by EHylands** [comment]
`but in the Apple Home App its pretty confusing group of switches and the input names did not carry over there although it looks like you got the labeling in the screen shot you sent me`

This is unfortunately a known issue with iOS 16. If you go in each switch accessory details and delete "Zone1 Inputs", you will find the intended input name as a placeholder. I hope the issue gets fixed in iOS 16.2.

`All that said I'm more interested in using the input selectors via Siri than the Home App UI`

I had good success with shortcuts and Siri to turn on the receiver with proper input setting. Using Siri directly on Home App accessories doesn't give an amazing result in my situation (many accessories with similar french name ...)

`It does appear they are mutually exclusive as must be the case since only one input at a time is selectalbe.`

All input are mutulally exclusive like stated and will all be turned off when receiver is powered off.

**Comment by EHylands** [comment]
Updated in https://github.com/EHylands/homebridge-anthemreceiver/pull/7

**Comment by aboulfad** [comment]
> Updated in #7

Thank you for the plugin, very useful for home AVR automation. Using latest version **0.7.0**, for some reason on **iOS16.5.1**, I get: `Not Supported` for `Zone1 Inputs`. I was expecting that I'd get individual switches corresponding to the discovered inputs.

![IMG_3525](https://github.com/EHylands/homebridge-anthemreceiver/assets/16640625/abf29b81-b9aa-4269-a41e-24a03bf889df)

I enabled HB DEBUG=*, but I can't find anything that explains this.
[anthem-receiver.log](https://github.com/EHylands/homebridge-anthemreceiver/files/11875439/anthem-receiver.log)

**Comment by EHylands** [comment]
Hi,
thank you for reporting this bug.
With MRX 740, the plugin is able to read all available inputs information from the receiver IP interface even if the zone is powered off. This allows the plugin to properly configure the Zone1 Inputs Accessory even while the receiver is powered off.

With MRX 720, only a few commands are available through the receiver IP interface while a Zone is powered off. At plugin start time, the plugin is unable to properly read the inputs name and fail to properly create the Zone1 Inputs accessory. This is why it is shown as not supported in Home App.

I will need to think about the best solution. It is a limitation with the available MRX *20 IP interface commands. Maybe the best option is to power on Zone1 if not on at plugin start time to be able to read all available information from receiver and properly configure all the accessoires.

**Comment by aboulfad** [comment]
Thank you for the explanation. How about looping over the # of inputs and using `ILNyy?` which works in standby! I thought `ICN?` would work to retrieve the # of inputs but it seems it isn't supported. Reference: [MRX x20 IP/serial codes](https://www.anthemav.com/downloads/MRX-x20-AVM-60-IP-RS-232.xls)

```
DEBUG=* node 4testmrx.js 
Connect: 1.786ms
  _MRX sendMsg ILN01?; +0ms
  _MRX sendMsg ILN02?; +1ms
  ...
  _MRX CONNECTED TO: 10.10.10.10:14999 +1ms
  _MRXDATA ILN01TV; +0ms
  _MRXDATA ILN02Apple TV;ILN03ATV 4K2;ILN04Oppo;ILN05Oppo COA;ILN06Play-Fi;ILN07PS5; +9ms
```

Merci ; )

**Comment by aboulfad** [comment]
Another suggestion for improvement to aid in automation. The x20 series (MRX720) is only ready to take commands after 6500ms from being powered up, e.g. change input. 
If one would want to have an automation to power the MRX and change input using the added external accessory, that would not work, as the plugin would send the power command followed right away by the input change whilst the MRX isn't ready. (pls see below where Z1INP4 is sent, but later Z1INP1 is read.)

Having individual switches for the inputs would provide a workaround where the user can add a "wait" command before sending the input change, or the plugin would have to account for this delay.

```
2023-06-27T11:16:32.307Z HAP-NodeJS:Accessory [MRX-720 EDE2] Processing characteristic set: {"characteristics":[{"aid":1,"iid":9,"value":1},{"aid":1,"iid":10,"value":4}]}
[27/06/2023, 07:16:32] [AnthemReceiver] Sending: Z1POW1;
[27/06/2023, 07:16:32] [AnthemReceiver] Sending: Z1INP4;
...
2023-06-27T11:16:32.314Z HAP-NodeJS:Accessory [MRX-720 EDE2] Setting Characteristic "Active" to value 1
2023-06-27T11:16:32.314Z HAP-NodeJS:Accessory [MRX-720 EDE2] Setting Characteristic "Active Identifier" to value 4
2023-06-27T11:16:32.317Z HAP-NodeJS:EventedHTTPServer:Connection [ipA] HTTP Response is finished
[27/06/2023, 07:16:32] [AnthemReceiver] Reading: "Z1POW1"
[27/06/2023, 07:16:32] [AnthemReceiver] Sending: Z1INP?;Z1MUT?;Z1VOL?;Z1SMD?;ICN?;Z1ALM?;
[27/06/2023, 07:16:32] [AnthemReceiver] Sending: 
[27/06/2023, 07:16:32] [AnthemReceiver] Reading: ""
2023-06-27T11:16:32.583Z HAP-NodeJS:EventedHTTPServer:Connection [ipA] Sending HAP event notifications [ { aid: 3, iid: 9, value: 1 } ]
[27/06/2023, 07:16:34] [AnthemReceiver] Reading: ""
[27/06/2023, 07:16:34] [AnthemReceiver] Reading: "Z1INP1"
...
```

**Comment by EHylands** [comment]
Hi,
As per MRX720 IP documentation:  

`Ony a few commands are valid when the system is in standby 'IDM?', 'ZxPOWy', 'ZxPOW?'. All other commands are considered 'Invalid'`

I think from previous test that ICN command is not available in standby mode.

Can you confirm that ISNyy? command works in standby mode ? 
What is the maximum number of inputs on X720 models ?
I could configure the inputs without relying on the ICN command.

When properly configured, the new Zone Inputs accessory contains individuals switch for every inputs.

The delay between the receiver reporting to be powered on `Reading: "Z1POW1"` and the receiver being able to accept command is a known issue. The receiver should only report `Reading: "Z1POW1"` when ready to accept command and not the other way around ...

**Comment by aboulfad** [comment]
Actually I was wrong, ICN also works in standby, I forgot the ";"  so ICN, ILN, ISN do work in standby on my MRX720 :) 

```
DEBUG=* node 6testmrx.js 
Connect: 1.678ms
  _MRX sendMsg ICN?; +0ms
  _MRX sendMsg ISN02?; +1ms
  _MRX sendMsg ILN02?; +0ms
  _MRX CONNECTED TO: 10.10.10.10:14999 +0ms
  _MRXDATA ICN9; +0ms
  _MRXDATA ISN02Apple TV;ILN02Apple TV; +9ms
```

**Comment by EHylands** [comment]
> Actually I was wrong, ICN also works in standby, I forgot the ";" so ICN, ILN, ISN do work in standby on my MRX720 :)
> 
> ```
> DEBUG=* node 6testmrx.js 
> Connect: 1.678ms
>   _MRX sendMsg ICN?; +0ms
>   _MRX sendMsg ISN02?; +1ms
>   _MRX sendMsg ILN02?; +0ms
>   _MRX CONNECTED TO: 10.10.10.10:14999 +0ms
>   _MRXDATA ICN9; +0ms
>   _MRXDATA ISN02Apple TV;ILN02Apple TV; +9ms
> ```

Great ! will be able to configure inputs at start time then.

**Comment by EHylands** [comment]
You can install alternate version 0.7.2
I enable input discovery at start time for MRX X20 model even if receiver is in standby mode.
You may have to remove and readd the input accessory to see changes.

**Comment by aboulfad** [comment]
Thank you - It does discover the inputs in standby, but the exposed switches for each input does not have the discovered input name even when you delete the Zone1 Inputs switch name.

Basically they all show `InputX undefined` instead of the corresponding discovered input name in the accessory detail view. I can provide a redacted full log with DEBUG=* if required.

PS: I have a test rpi running HB, so I wipe out clear all accessories and bridge from HomeKit before I did that test.

**Comment by EHylands** [comment]
I did minor changes in 0.7.3 beta . Not sure that was the problem though.
I you want to provide debut information, you can only run homebride -D and provide with the log file.
Tnx

**Comment by aboulfad** [comment]
sorry, still shows `InputX undefined`. I guess accessory name isn't being updated.
[v0.7.3.log](https://github.com/EHylands/homebridge-anthemreceiver/files/11919150/v0.7.3.log)

**Comment by EHylands** [comment]
Found the error !
You can run 0.7.4
Plugin will wait for all inputs to be discovered before starting operation !

**Comment by EHylands** [comment]
@aboulfad , were you able to run the inputs accessory on MRX720 with last version ?

**Comment by aboulfad** [comment]
> @aboulfad , were you able to run the inputs accessory on MRX720 with last version ?

Oh sorry, when I liked your previous post, it was to ACK that I did, thank you!
